cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/runtimejob/TetheringRuntimeJobManager.java

```diff
@@ -132,4 +143,54 @@ void publishToControlChannel(TetheringControlMessage message) throws Exception {
       throw new Exception(String.format("Failed to publish to topic %s", topicId), e);
     }
   }
+
+  /**
+   * Use GZIPOutputStream to compress the LocalFile and Base64 to encode as String
+   */
+  @VisibleForTesting
+  String getLocalFileAsCompressedString(LocalFile localFile) throws IOException {
+    File file = new File(localFile.getURI());
+    byte[] buffer = new byte[1024 * 500]; // use 500kb buffer
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    try (GZIPOutputStream os = new GZIPOutputStream(baos); FileInputStream fis = new FileInputStream(file)) {
+      int length;
+      while ((length = fis.read(buffer)) > 0) {
+        os.write(buffer, 0, length);
+      }
+    }
+    return Base64.encodeBase64String(baos.toByteArray());
+  }
+
+  /**
+   * Add select LocalFiles and cConf entries to the control message payload
+   */
+  private TetheringLaunchMessage createLaunchPayload(Collection<? extends LocalFile> localFiles) throws IOException {
+    Map<String, String> localizeFiles = new HashMap<>();
+    Map<String, String> cConfEntries = new HashMap<>();
+
+    LocalFile logbackFile = localFiles.stream()
+      .filter(file -> file.getName().equals(DistributedProgramRunner.LOGBACK_FILE_NAME))
+      .findFirst()
+      .orElseThrow(() -> new IllegalStateException("Cannot find logback file"));
+    localizeFiles.put(DistributedProgramRunner.LOGBACK_FILE_NAME, getLocalFileAsCompressedString(logbackFile));
+
+    LocalFile programOptionsFile = localFiles.stream()
+      .filter(file -> file.getName().equals(DistributedProgramRunner.PROGRAM_OPTIONS_FILE_NAME))
+      .findFirst()
+      .orElseThrow(() -> new IllegalStateException("Cannot find program options file"));
+    localizeFiles.put(DistributedProgramRunner.PROGRAM_OPTIONS_FILE_NAME,
+                      getLocalFileAsCompressedString(programOptionsFile));
+
+    LocalFile appSpecFile = localFiles.stream()
+      .filter(file -> file.getName().equals(DistributedProgramRunner.APP_SPEC_FILE_NAME))
+      .findFirst()
+      .orElseThrow(() -> new IllegalStateException("Cannot find app spec file"));
+    localizeFiles.put(DistributedProgramRunner.APP_SPEC_FILE_NAME, getLocalFileAsCompressedString(appSpecFile));
+
+    for (String prefix : SELECT_CCONF_FIELDS) {
+      String prefixRegex = "^" + prefix.replace(".", "\\.");
+      cConfEntries.putAll(cConf.getValByRegex(prefixRegex));
+    }
```

Comment:
Should we generate a run id here and pass it along in the TetheringLaunchMessage?